imod/msw/coupler_mapping.py

```diff
@@ -35,59 +28,63 @@ class CouplerMapping(MetaSwapPackage):
-
-    def _create_mod_id_rch(self, svat):
+    def _create_mod_id_rch(
+        self, svat: xr.DataArray, idomain_top_active: xr.DataArray
+    ) -> xr.DataArray:
         """
         Create modflow indices for the recharge layer, which is where
         infiltration will take place.
         """
-        self.dataset["mod_id"] = xr.full_like(svat, fill_value=0, dtype=np.int64)
+        mod_id = xr.full_like(svat, fill_value=0, dtype=np.int64)
         n_subunit = svat["subunit"].size
-        idomain_top_active = self.idomain_active.sel(layer=1, drop=True)
 
         n_mod_top = idomain_top_active.sum()
 
         # idomain does not have a subunit dimension, so tile for n_subunits
-        mod_id_1d = np.tile(np.arange(1, n_mod_top + 1), (n_subunit, 1))
+        mod_id_1d: IntArray = np.tile(np.arange(1, n_mod_top + 1), (n_subunit, 1))
 
-        self.dataset["mod_id"].values[:, idomain_top_active.values] = mod_id_1d
+        mod_id.data[:, idomain_top_active.data] = mod_id_1d
+        return mod_id
+
+    def _render(
+        self,
+        file: TextIO,
+        index: IntArray,
+        svat: xr.DataArray,
+        mf6_dis: StructuredDiscretization,
+        mf6_well: Mf6Wel,
+        *args: Any,
+    ):
+        if mf6_well and (not isinstance(mf6_well, Mf6Wel)):
+            raise TypeError(rf"mf6_well not of type 'Mf6Wel', got '{type(mf6_well)}'")
+        # Test if equal or larger than 1, to ignore idomain == -1 as well. Don't
+        # assign to self.dataset, as grid extent might differ from svat when
+        # MetaSWAP only covers part of the Modflow grid domain.
+        idomain_active = mf6_dis["idomain"] >= 1
+        idomain_top_active = idomain_active.sel(layer=1, drop=True)
+        mod_id = self._create_mod_id_rch(svat, idomain_top_active)
 
-    def _render(self, file, index, svat: xr.DataArray):
-        self._create_mod_id_rch(svat)
         # package check only possible after calling _create_mod_id_rch
         self._pkgcheck()
 
-        data_dict = {"svat": svat.values.ravel()[index]}
-
+        data_dict: DataDictType = {"svat": svat.values.ravel()[index]}
         data_dict["layer"] = np.full_like(data_dict["svat"], 1)
-
-        for var in self._with_subunit:
-            data_dict[var] = self._index_da(self.dataset[var], index)
+        data_dict["mod_id"] = self._index_da(mod_id, index)
 
         # Get well values
-        if self.well:
-            mod_id_well, svat_well, layer_well = self._create_well_id(svat)
-            data_dict["mod_id"] = np.append(mod_id_well, data_dict["mod_id"])
-            data_dict["svat"] = np.append(svat_well, data_dict["svat"])
-            data_dict["layer"] = np.append(layer_well, data_dict["layer"])
+        if mf6_well:
+            well_data_dict = self._create_well_id(svat, idomain_active, mf6_well)
+            for key in data_dict.keys():
+                data_dict[key] = np.append(well_data_dict[key], data_dict[key])
```

Comment:
That is also a good way of doing things. It has the slight advantage that there is an additional variable being named, which might improve reality a bit, though updating keys in a for loop is such a basic thing in python that most developers will not struggle with this part of the code.

The slight advantage of the present approach is a slightly reduced memory use, as variables are updated one by one, instead of creating a full dictionary ``appended_well_data`` first and updating the existing dict with it. Though I do not think in most cases memory use here will matter a lot.

I think I'll stick to the present approach, it is a bit of a micro-optimalization; I do not have a strong opinion on the best approach, and I don't think it will matter a lot in the end.